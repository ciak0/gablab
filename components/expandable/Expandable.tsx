import {
  FunctionComponent, ReactChild, useCallback, useState,
} from 'react';

export interface ExpandableProps {
  summary: ReactChild,
  expanded?: boolean,
  className?: string,
  expandText?: string,
  collapseText?: string,
}

const Expandable: FunctionComponent<ExpandableProps> = ({
  summary,
  expanded,
  className,
  children,
  expandText,
  collapseText,
}) => {
  const [isExpanded, setExpanded] = useState(expanded ?? false);
  const toggle = useCallback(() => {
    setExpanded(!isExpanded);
  }, [isExpanded]);

  return (
    <div className={className}>
      {summary}
      {!isExpanded && children && (
        <button
          className="block mt-4 text-sm text-neutral-500"
          type="button"
          onClick={toggle}
        >
          {expandText}
        </button>
      )}
      {isExpanded && children}
      {isExpanded && (
        <button
          className="block mt-4 text-sm text-neutral-500"
          type="button"
          onClick={toggle}
        >
          {collapseText}
        </button>
      )}
    </div>
  );
};

Expandable.defaultProps = {
  className: '',
  expandText: '...see more',
  collapseText: 'see less',
};

export default Expandable;
