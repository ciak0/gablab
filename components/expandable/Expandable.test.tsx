import React from 'react';
import { render, screen } from '@testing-library/react';
import Expandable from './Expandable';

describe('components/Expandable', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Expandable
        summary=""
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the summary and not the children', () => {
    render(
      <Expandable
        summary="Summary"
      >
        <div>Children</div>
      </Expandable>,
    );

    expect(screen.getByText('Summary')).toBeInTheDocument();
    expect(screen.queryByText('Children')).toBeNull();
  });

  it('should render the children, when the expand button is clicked, and remove them when the collapse is clicked', () => {
    render(
      <Expandable
        summary="Summary"
        expandText="see more"
        collapseText="see less"
      >
        <div>Children</div>
      </Expandable>,
    );

    screen.getByText('see more').click();
    expect(screen.getByText('Children')).toBeInTheDocument();

    screen.getByText('see less').click();
    expect(screen.queryByText('Children')).toBeNull();
  });
});
