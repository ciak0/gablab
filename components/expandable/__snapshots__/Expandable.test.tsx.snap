// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Expandable should render correctly 1`] = `
<div>
  <div
    class=""
  >
    
    <button
      class="block mt-4 text-sm text-neutral-500"
      type="button"
    >
      ...see more
    </button>
  </div>
</div>
`;
